<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Mendelaimselection.jl : This analysis option selects the SNPs that are most informative at predicting ancestry for your data — the best Ancestry Informative Markers (AIMs).">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Mendelaimselection.jl</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/OpenMendel/MendelAimSelection.jl">View on GitHub</a>

          <h1 id="project_title">Mendelaimselection.jl</h1>
          <h2 id="project_tagline">This analysis option selects the SNPs that are most informative at predicting ancestry for your data — the best Ancestry Informative Markers (AIMs).</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/OpenMendel/MendelAimSelection.jl/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/OpenMendel/MendelAimSelection.jl/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Overview</h3>

<p>Mendel AIM Selection is a component of the umbrella <a href="https://openmendel.github.io">OpenMendel</a> project. This analysis option selects the SNPs that are most informative at predicting ancestry for your data — the best Ancestry Informative Markers (AIMs).</p>

<h3>
<a id="appropriate-problems-and-data-sets" class="anchor" href="#appropriate-problems-and-data-sets" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Appropriate Problems and Data Sets</h3>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p><em>Note: Three OpenMendel packages - <a href="https://github.com/OpenMendel/SnpArrays.jl">SnpArrays</a>, <a href="https://github.com/OpenMendel/Search.jl">Search</a>, and <a href="https://github.com/OpenMendel/MendelBase.jl">MendelBase</a> must be installed before any Mendel analysis packages will run.</em></p>

<p>Within Julia, use the package manager to install MendelAIMSelection:</p>

<pre><code>Pkg.clone("git@github.com:OpenMendel/MendelAIMSelection.jl.git")
</code></pre>

<p>This package supports Julia v0.4.</p>

<h3>
<a id="input-files" class="anchor" href="#input-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input Files</h3>

<p>The MendelAIMSelection analysis package uses the following input files. Example input files can be found in the <a href="https://github.com/OpenMendel/MendelAIMSelection.jl/tree/master/docs">docs</a> subfolder of the MendelAIMSelection project. (An analysis won't always need every file type below.)</p>

<ul>
<li>
<a href="#control-file">Control File</a>: Specifies the names of your data input and output files and any optional parameters (<em>keywords</em>) for the analysis. (For a list of common keywords, see <a href="https://openmendel.github.io/MendelBase.jl/#keywords-table">Keywords Table</a>).</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#locus-file">Locus File</a>: Names and describes the genetic loci in your data.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#pedigree-file">Pedigree File</a>: Gives information about your individuals, such as name, sex, family structure, and ancestry.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#phenotype-file">Phenotype File</a>: Lists the available phenotypes.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#snp-definition-file">SNP Definition File</a>: Defines your SNPs with information such as SNP name, chromosome, position, allele names, allele frequencies.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#snp-data-file">SNP Data File</a>: Holds the genotypes for your data set. Must be a standard binary PLINK BED file in SNP major format. If you have a SNP data file you must have a SNP definition file.</li>
</ul>

<h3>
<a id="control-file" class="anchor" href="#control-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Control file<a id="control-file"></a>
</h3>

<p>The Control file is a text file consisting of keywords and their assigned values. The format of the Control file is:</p>

<pre><code>Keyword = Keyword_Value(s)
</code></pre>

<p>Below is an example of a simple Control file to run AIM Selection:</p>

<pre><code>#
# Input and Output files.
#
snpdata_file = 1000genomes_chr1_eas.bed
snpdefinition_file = 1000genomes_chr1_eas.snp
pedigree_file = 1000genomes_chr1_eas.ped
output_file = 1000genomes_chr1_eas Output.txt
#
# Analysis parameters for AIM Selection option.
#
</code></pre>

<p>In the example above, the only six keywords specify the input and output files: <em>1000genomes_chr1_eas.bed</em>, <em>1000genomes_chr1_eas.snp</em>, <em>1000genomes_chr1_eas.ped</em> and <em>1000genomes_chr1_eas Output.txt</em>. The text after the '=' are the keyword values. The names of keywords are <em>not</em> case sensitive. (The keyword values <em>may</em> be case sensitive.) A list of OpenMendel keywords common to most analysis package can be found <a href="https://openmendel.github.io/MendelBase.jl/#keywords-table">here</a>.</p>

<h3>
<a id="data-files" class="anchor" href="#data-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data Files</h3>

<p>AIM Selection requires a <a href="https://openmendel.github.io/MendelBase.jl/#control-file">Control file</a>, and a <a href="https://openmendel.github.io/MendelBase.jl/#pedigree-file">Pedigree file</a>. Genotype data can be included in the Pedigree file, in which case a <a href="https://openmendel.github.io/MendelBase.jl/#locus-file">Locus file</a> is required. Alternatively, genotype data can be provided in a <a href="https://openmendel.github.io/MendelBase.jl/#snp-data-file">SNP data file</a>, in which case a <a href="https://openmendel.github.io/MendelBase.jl/#snp-definition-file">SNP Definition File</a> is required. Details on the format and contents of the Control and data files can be found on the <a href="https://openmendel.github.io/MendelBase.jl">MendelBase</a> documentation page. There are example data files in the AIM Selection <a href="https://github.com/OpenMendel/MendelAIMSelection.jl/tree/master/docs">docs</a> folder.</p>

<h3>
<a id="running-the-analysis" class="anchor" href="#running-the-analysis" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Running the Analysis</h3>

<p>To run this analysis package, first launch Julia. Then load the package with the command:</p>

<pre><code> julia&gt; using MendelAIMSelection
</code></pre>

<p>Next, if necessary, change to the directory containing your files, for example,</p>

<pre><code> julia&gt; cd("~/path/to/data/files/")
</code></pre>

<p>Finally, to run the analysis using the parameters in the control file Control_file.txt use the command:</p>

<pre><code> julia&gt; AIMSelection("Control_file.txt")
</code></pre>

<p><em>Note: The package is called</em> MendelAIMSelection <em>but the analysis function is called simply</em> AIMSelection.</p>

<h3>
<a id="interpreting-the-results" class="anchor" href="#interpreting-the-results" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Interpreting the results</h3>

<h3>
<a id="citation" class="anchor" href="#citation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Citation</h3>

<p>If you use this analysis package in your research, please cite the following reference in the resulting publications:</p>

<p><em>Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.</em></p>



<h3>
<a id="acknowledgments" class="anchor" href="#acknowledgments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Acknowledgments</h3>

<p>This project is supported by the National Institutes of Health under NIGMS awards R01GM053275 and R25GM103774 and NHGRI award R01HG006139.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Mendelaimselection.jl maintained by <a href="https://github.com/OpenMendel">OpenMendel</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
